### PR TITLE
Fixed same enterprise appear several times in the fee dropdown list

### LINKED ISSUE #9603: [Admin][Order cycle] Same enterprise appear several times in the fees dropdown list
## Description
When choosing to apply fees in steps 3 and 4 of a hub's order cycle, the same enterprise can appear several time in the dropdown list.

## Expected behavior
One enterprise should appear only one time.

## Steps to Reproduce
1. Connect to an account that has a producer-hub enterprise (with at least one available product so it can be added as a supplier): for ex Enterprise 3949 and 4001
2. Create or edit an order cycle
3. At step 2, choose the hub as the supplier, then click "Add fee"
4. See that the hub appears several times

## Animated Gif/Screenshot
Screen Shot 2022-08-29 at 19 26 54

Screen Shot 2022-09-19 at 11 32 43

## Workaround
Choose whether of the proposed options.

## Severity
bug-s4: it's annoying, but you can use it
0

### PR DESCRIPTION
**Pull request by maniSHarma7575**
#### What? Why?
When choosing to apply fees in steps 3 and 4 of a hub's order cycle, the same enterprises was appearing several time in the drop-down list.
Closes #9603 

**Before:**
Screenshot 2022-10-08 at 6 15 56 PM
Screenshot 2022-10-08 at 6 16 26 PM

**After:**

Screenshot 2022-10-08 at 6 17 03 PM
Screenshot 2022-10-08 at 6 40 27 PM

#### What should we test?

 - Connect to an account that has a producer-hub enterprise (with at least one available product so it can be added as a supplier)
 - Create or edit an order cycle
 - At step 2, choose the hub as the supplier, then click "Add fee"
 - See that the hub should not appears several times

#### Release notes

Changelog Category: Admin facing changes | Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by maniSHarma7575** [comment]
@audez @sigmundpetersen PR is ready for review.

**Review by mkllnk — APPROVED** [review]
Nice, thank you.

We may get rid of jQuery one day but we can update this code then. Great work.

File: app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee [hunk 1]
```diff
@@ -18,7 +18,7 @@ angular.module('admin.orderCycles')
       OrderCycle.exchangeDirection(exchange)
 
     $scope.enterprisesWithFees = ->
-      $scope.enterprises[id] for id in [OrderCycle.participatingEnterpriseIds()..., [OrderCycle.order_cycle.coordinator_id]...] when $scope.enterpriseFeesForEnterprise(id).length > 0
+      $scope.enterprises[id] for id in jQuery.unique([OrderCycle.participatingEnterpriseIds()..., [OrderCycle.order_cycle.coordinator_id]...].sort()) when $scope.enterpriseFeesForEnterprise(id).length > 0
```
Comment (jibees) [inline app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee:21]: I'd prefer to use jQuery.

Why not using the `Set` type: 

```diff
diff --git a/app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee b/app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee
index 185d5009d..ee7c1aa5c 100644
--- a/app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee
+++ b/app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee
@@ -18,7 +18,8 @@ angular.module('admin.orderCycles')
       OrderCycle.exchangeDirection(exchange)
 
     $scope.enterprisesWithFees = ->
-      $scope.enterprises[id] for id in jQuery.unique([OrderCycle.participatingEnterpriseIds()..., [OrderCycle.order_cycle.coordinator_id]...].sort()) when $scope.enterpriseFeesForEnterprise(id).length > 0
+      ids = [OrderCycle.participatingEnterpriseIds()..., [OrderCycle.order_cycle.coordinator_id]...]
+      $scope.enterprises[id] for id in Array.from(new Set(ids)) when $scope.enterpriseFeesForEnterprise(id).length > 0
 
     $scope.removeExchange = ($event, exchange) ->
       $event.preventDefault()
```

File: app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee [hunk 2]
```diff
@@ -93,8 +93,8 @@ angular.module('admin.orderCycles').factory 'OrderCycle', ($resource, $window, $
       variant_ids
 
     participatingEnterpriseIds: ->
-      suppliers = (exchange.enterprise_id for exchange in this.order_cycle.incoming_exchanges)
-      distributors = (exchange.enterprise_id for exchange in this.order_cycle.outgoing_exchanges)
+      suppliers = (parseInt(exchange.enterprise_id) for exchange in this.order_cycle.incoming_exchanges)
+      distributors = (parseInt(exchange.enterprise_id) for exchange in this.order_cycle.outgoing_exchanges)
       jQuery.unique(suppliers.concat(distributors)).sort()
```
Comment (jibees) [inline app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee:98]: Probably out of scope, but as we are modifying the file...

```diff
diff --git a/app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee b/app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee
index 10c624975..88ad9da1b 100644
--- a/app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee
+++ b/app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee
@@ -95,7 +95,7 @@ angular.module('admin.orderCycles').factory 'OrderCycle', ($resource, $window, $
     participatingEnterpriseIds: ->
       suppliers = (parseInt(exchange.enterprise_id) for exchange in this.order_cycle.incoming_exchanges)
       distributors = (parseInt(exchange.enterprise_id) for exchange in this.order_cycle.outgoing_exchanges)
-      jQuery.unique(suppliers.concat(distributors)).sort()
+      Array.from(new Set([suppliers..., distributors...]))
 
     exchangesByDirection: (direction) ->
       if direction == 'incoming'
```

File: app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee (hunk 1 above)
Comment (jibees) [reply app/assets/javascripts/admin/order_cycles/controllers/order_cycle_exchanges_controller.js.coffee:21]: I'm also wondering the role of the `sort()` method. Why would we sort by id?

File: app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee (hunk 2 above)
Comment (jibees) [reply app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee:98]: (same question here around the `sort()` method ...)

**Review by jibees — CHANGES_REQUESTED** [review]
Thanks for this one!

I think I'd prefer that not use jQuery, and that's why I've made a proposal with `Set()`. I'm ok with a `filter()` method on the array also! (something like `list.filter((x, i, a) => a.indexOf(x) === i)`)

**Comment by maniSHarma7575** [comment]
@jibees Feedbacks done!

**Review by jibees — APPROVED** [review]
Nice, thanks!

**Comment by drummer83** [comment]
Hi @maniSHarma7575,
Thank you for this pull request! :fire: 

I have tested the behavior before and after staging your pull request.

**Before:**
Peek 2022-10-10 17-45

**After:**
Peek 2022-10-10 18-12

As you can see, the enterprise is only appearing once with you pull request! Awesome!
I did not find any problems around this solution. This is ready to go! :rocket: 

Thanks again for this great contribution!

**Comment by maniSHarma7575** [comment]
Thanks @drummer83 @mkllnk @jibees For merging this. Happy to be part of it!

**Pull request merged**